traits/trait_list_object.py

```diff
@@ -53,460 +88,755 @@ def __init__(self, index=0, removed=None, added=None):
-        The HasTraits instance that the list has been set as an attribute for.
-    name : str
-        The name of the attribute on the object.
     value : list
-        The list of values to initialize the TraitListObject with.
+        The value for the list
+    validator : callable
+        Called to validate items in the list
+    notifiers : list of callable
+        A list of callables with the signature::
+
+         notifier(trait_list, index, removed, added)
 
     Attributes
     ----------
-    trait : CTrait instance
-        The CTrait instance associated with the attribute that this list
-        has been set to.
-    object : weak reference to a HasTraits instance
-        A weak reference to a HasTraits instance that the list has been set
-        as an attribute for.
-    name : str
-        The name of the attribute on the object.
-    name_items : str
-        The name of the items event trait that the trait list will fire when
-        mutated.
+    value : list
+        The value for the list
+    validator : callable
+        Called to validate items in the list
+    notifiers : list of callable
+        A list of callables with the signature::
+
+         notifier(trait_list, index, removed, added)
+
     """
 
-    def __init__(self, trait, object, name, value):
-        self.trait = trait
-        self.object = ref(object)
-        self.name = name
-        self.name_items = None
-        if trait.has_items:
-            self.name_items = name + "_items"
+    # ------------------------------------------------------------------------
+    # TraitList interface
+    # ------------------------------------------------------------------------
+
+    def validate(self, index, removed, added):
+        """ Validate values for given index and removed values.
 
-        # Do the validated 'setslice' assignment without raising an
-        # 'items_changed' event:
-        if trait.minlen <= len(value) <= trait.maxlen:
-            try:
-                validate = trait.item_trait.handler.validate
-                if validate is not None:
-                    value = [validate(object, name, val) for val in value]
+        This simply calls the validator provided by the class, if any.
+        The validator is expected to have the signature::
 
-                list.__setitem__(self, slice(0, 0), value)
+            validator(trait_list, index, removed, added)
 
-                return
+        and return a validated added value or list of added values
+        or raise a TraitError.
 
-            except TraitError as excp:
-                excp.set_prefix("Each element of the")
-                raise excp
+        Parameters
+        ----------
+        index : int or slice
+            The indices being modified by the operation.
+        removed : object or list
+            The item or items to be removed.
+        added : object or iterable
+            The new item or items being added to the list.
```

Comment:
This is a comment from previous, but GitHub hid it.
Could we use a list for both `removed` and `added` **please**?
It would be impossible to handle list of list of ... of list for EEP 3 if this `removed` and `added` can be one item or many items. The listener has no context as for whether a change is invoked via `append` or via `extend`, it just knows the content of the list has changed somehow.

Reply:
I've been working on this, so there are a couple of cases that need to be considered. 
The problem is differentiating a single object from an actual list, For instance consider setting a list at some index `lst[0] = [5]`, note the `index` in this case is not a slice, so is this going to be sent as `[[5]]` to the validator ? We would also have to pass in the information to differentiate an item wrapped in a list vs an actual list.

Reply:
There's also consistency with the `TraitListEvent` to consider: there `added` and `removed` are always lists of items, and it would be confusing if that were different here. In general, APIs where a parameter or return value can have multiple different types are awkward to work with.

Reply:
Changed added and removed to always be a list.